### PR TITLE
Created Operators for Bing Maps

### PR DESCRIPTION
**Pull request by lucas-sio-rosa**
Operators for Bing Maps created:
- CreateJob - takes a PIPE file from GCS and uploads it to the BingMaps endpoint to create a Geocode Job
- JobStatus - a sensor that GETs the job id endpoint to check if the geocoding is finished
- DownloadJob - downloads a completed job to a local file and uploads it to GCS

### CODE REVIEW

File: debussy/hooks/bing.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    def is_downloadable(response):
+        """Checks if the response contains a downloadable file, by verifiying the content-type.
+        A json and html types are ignored and anything else (should only be text) is downloadable
+        """
+        header = response.headers
+        content_type = header.get('content-type')
+        if 'json' in content_type.lower():
+            return False
+        if 'html' in content_type.lower():
+            return False
```
Comment (embatbr) [inline debussy/hooks/bing.py]: Replace it by
```python
if content_type.lower() in ('json', 'html'):
    return False
```

File: debussy/hooks/bing.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        :type operation: str
+        :param file_path: when posting, indicates the location of the local file to be sent.
+            When getting, indicates where the file should be written to.
+            Can be ignored when checking status
+        :type file_path: str
+        """
+        url = '{}/{}'.format(self.conn.host, method)
+        api_params['key'] = self.conn.password
+        
+        if(operation == 'GET'):
```
Comment (embatbr) [inline debussy/hooks/bing.py]: No need for parentheses.

File: debussy/hooks/bing.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            response = requests.get(url, params=api_params)
+
+            if(self.is_downloadable(response)):
```
Comment (embatbr) [inline debussy/hooks/bing.py]: No need for parentheses.

File: debussy/hooks/bing.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                open(file_path, 'wb').write(response.content)
+        elif(operation == 'POST'):
```
Comment (embatbr) [inline debussy/hooks/bing.py]: No need for parentheses.

File: debussy/hooks/bing.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            with open(file_path,'r') as r:
+                content = r.read()
+
+            response = requests.post(url, data=content.encode('utf-8'), params=api_params, headers={'Content-Type': 'text/plain'})
+        else:
+            raise AirflowException('Currently unsupported call operation: {}'.format(operation))
+        
+        # when a file will be downloaded, we simply output the file_path (to prevent the entire document to be outputted)
+        # else, we can safely output the response json
+        logging.info('File: {}'.format(file_path) if(self.is_downloadable(response)) else response.json())
```
Comment (embatbr) [inline debussy/hooks/bing.py]: Is this a ternary if? Wouldn't it be better to use regular if-else formatting?

File: debussy/hooks/bing.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    The hook expects an HTTP Airflow connection to be created, with the host set as "http://spatial.virtualearth.net/REST/v1/dataflows/geocode",
+    and the password set as the user's Bing Maps key. The hook supports calling endpoints as GET (to check status and download results) and as POST (to create jobs).
+    .. seealso::
+        Follow a walkthrough of the Bing Maps Geocode Batch Job at:
+        https://docs.microsoft.com/en-us/bingmaps/spatial-data-services/geocode-dataflow-api/geocode-dataflow-walkthrough
+    
+    :param bing_maps_conn_id: name of the HTPP connection with the Bing Geocode endpoint and Key
+    :type bing_maps_conn_id: str
+    """
+    def __init__(self, bing_maps_conn_id='bing_maps_default'):
```
Comment (embatbr) [inline debussy/hooks/bing.py]: Shouldn't you call the BaseHook __init__ operator?

File: debussy/operators/bigquery.py [hunk 7]
```diff
@@ -2,29 +2,73 @@
...
+from airflow.utils.decorators import apply_defaults
 from airflow.contrib.hooks.gcs_hook import GoogleCloudStorageHook
+from airflow.contrib.hooks.bigquery_hook import BigQueryHook
 from airflow.contrib.operators.bigquery_operator import BigQueryOperator
 
+from airflow import AirflowException
+from airflow.models import BaseOperator
+
+class BigQueryDropTableOperator(BaseOperator):
+    """Classe que força o BigQuery a deletar uma tabela.
```
Comment (embatbr) [inline debussy/operators/bigquery.py]: English, motherfucker, do you speak it?!

File: debussy/operators/bigquery.py [hunk 8]
```diff
@@ -132,3 +176,109 @@ def execute(self, context):
...
+        # preparing the update clause
+        # the helper function allows us to sequentially traverse the fields
+        # including nested record to filter out any exceptions that came with the ignore argument
+        update_clause = ''
+        for f, l in json_traverser(table_schema):
+            # constructing a complete path so that we can ignore specific fields or subfields
+            full_field_name = f['name'] if l == '' else '{}.{}'.format(l, f['name'])
+            
+            # ignore whole records, specific fields/subfields, pk fields
+            if(l not in update_fields_ignore and full_field_name not in update_fields_ignore and full_field_name not in pk_fields):
```
Comment (embatbr) [inline debussy/operators/bigquery.py]: No need for parentheses.

File: debussy/operators/bigquery.py [hunk 9]
```diff
@@ -132,3 +176,109 @@ def execute(self, context):
...
+                update_clause += 'd.{0} = s.{0},\n'.format(full_field_name)
+        update_clause = update_clause[:-2]
+
+        # preparing the insert clause
+        # only whole records or root fields can be referenced here
+        # so a simple pass through the list is enough
+        insert_list = ''
+        for f in table_schema:
+            if(f['name'] not in insert_fields_ignore):
```
Comment (embatbr) [inline debussy/operators/bigquery.py]: No need for parentheses.

File: debussy/operators/bing.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        bm_hook = BingMapsHook(bing_maps_conn_id=self.bing_maps_conn_id)
+
+        method = '{}/output/succeeded'.format(job_id)
+        bm_hook.call(method=method, api_params={}, operation='GET', file_path=local_file_path)
+
+        # processing the file and uploading to the bucket
+        with open(local_file_path, 'r') as rf:
+            with open(prepared_file_path, 'w') as wf:
+                for num, line in enumerate(rf, 1):
+                    if(num == 1):
```
Comment (embatbr) [inline debussy/operators/bing.py]: No need for parentheses.

File: debussy/operators/bing.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+                        pass
+                    elif(num == 2):
```
Comment (embatbr) [inline debussy/operators/bing.py]: No need for parentheses.

File: debussy/operators/python.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
+# -*- coding: utf-8 -*-
+
+from datetime import datetime
+
+from airflow.operators.python_operator import BranchPythonOperator
+
+def checa_dia_mes(ds, **kwargs):
+    date = datetime.strptime(ds, '%Y-%m-%d')
+
+    if(date.day == kwargs['dia']):
```
Comment (embatbr) [inline debussy/operators/python.py]: No need for parentheses.

File: debussy/sensors/bing.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+        # calling the endpoint to get the job status
+        bm_hook = BingMapsHook(bing_maps_conn_id=self.bing_maps_conn_id)
+        response = bm_hook.call(method=job_id, operation='GET', api_params={'output': 'json'})
+        
+        # if the response was successfull
+        if response:
+            # check the status of the job - Aborted is a failure case, pending is waiting and Completed is a success
+            status = response.json()['resourceSets'][0]['resources'][0]['status']
+            if(status == 'Completed'):
```
Comment (embatbr) [inline debussy/sensors/bing.py]: No need for parentheses.

File: debussy/sensors/bing.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+                return True
+            elif(status == 'Pending'):
```
Comment (embatbr) [inline debussy/sensors/bing.py]: No need for parentheses.

**Comment by embatbr** [comment]
Comment geral: coloca uma linha vazia no final dos arquivos.

File: debussy/hooks/bing.py (hunk 6 above)
Comment (lucas-sio-rosa) [reply debussy/hooks/bing.py]: I used the SlackHook implementation as my basis and they don't call the BaseHook init, so I assumed it to be unnecessary

**Pull request merged**